File: src/main/java/com/powa/gtin/GTIN.java
/**
 * Copyright (C) 2015 Powa Technologies Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.powa.gtin;

/**
 * Valid GTIN code. Static methods are provided for identifying, validating and parsing GTIN codes
 * represented as strings.
 */
public final class GTIN {

    private final String gtin;
    private final GTINFormat format;

    // Prevent instantiation
    private GTIN(final String gtin) {
        this.gtin = gtin;
        format = GTINFormat.forLength(gtin.length());
    }

    /**
     * Parses a GTIN string into a GTIN object.
     * You may use the {@link #isValid(String)} method first if you are unsure whether the
     * string is a valid GTIN to avoid.
     *
     * @param gtin the GTIN string.
     * @return a GTIN object if the string is a valid GTIN.
     * @throws InvalidGTINException if the string is not a valid GTIN.
     */
    public static GTIN parse(final String gtin) throws InvalidGTINException {
        if (!isValid(gtin)) {
            throw new InvalidGTINException("Cannot parse invalid string " + gtin + " to string");
        }

        return new GTIN(gtin);
    }

    /**
     * Checks the GTIN for validity by checking the format and checksum.
     *
     * @param gtin the GTIN, may be GTIN-8, GTIN-12, GTIN-13 or GTIN-14 format.
     * @return {@code true} if the GTIN is valid, {@code false} otherwise.
     */
    public static boolean isValid(final String gtin) {
        // Check format of barcode for validity
        if (!matchesFormat(gtin)) {
            return false;
        }

        int checkSum = 0;
        int gtinLength = gtin.length();
        int weightBit = gtinLength % 2;
        for (int i = 0; i < gtinLength; i++) {
            int weight = i % 2 == weightBit ? 3 : 1;
            checkSum += Integer.parseInt(gtin.substring(i, i + 1)) * weight;
        }

        return checkSum % 10 == 0;
    }

    /**
     * Checks whether the input string matches the GTIN format, i.e. is of the correct length for
     * a GTIN and that the string contains only digits.
     *
     * @param gtin the possible GTIN string.
     * @return {@code true} if the input string is a valid GTIN string, {@code false} otherwise.
     */
    public static boolean matchesFormat(final String gtin) {
        if (gtin == null) {
            throw new IllegalArgumentException("gtin is null");
        }

        // Check length is ok
        int gtinLength = gtin.length();
        if (gtinLength != 8 && gtinLength != 12 && gtinLength != 13 && gtinLength != 14) {
            return false;
        }

        // Check whether is a number
        return gtin.matches("^\\d+$");
    }

    /**
     * @return the format or type of this GTIN.
     */
    public GTINFormat format() {
        return format;
    }

    /**
     * @return the length of this GTIN.
     */
    public int length() {
        return format.length();
    }

    /**
     * @return this GTIN code as a string.
     */
    @Override
    public String toString() {
        return gtin;
    }

}

File: gtin.objc/GTINObjc/GTIN.m
/**
 * Copyright (C) 2015 Powa Technologies Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "GTIN.h"

@interface GTIN()

@property (nonatomic, strong, readwrite) NSString *gtin;
// Redeclare as readwrite
@property (nonatomic, readwrite) GTINFormat format;

@end

@implementation GTIN

#define kGtinKey       @"gtin"

- (id)initWithCoder:(NSCoder *)decoder {
    NSString *gtin = [decoder decodeObjectForKey:kGtinKey];
    
    GTIN *g = [GTIN new];
    g.gtin = gtin;
    g.format = GTINFormatForLength(gtin.length);
    
    return g;
}

+ (GTIN *)gtin:(NSString *)gtin {
    if (![GTIN isValidString:gtin]) {
        @throw [NSException exceptionWithName:GTINFormatException reason:[NSString stringWithFormat:@"String %@ is not a valid gtin", gtin] userInfo:nil];
    }
    
    GTIN *g = [GTIN new];
    g.gtin = gtin;
    g.format = GTINFormatForLength(gtin.length);
    
    return g;
}

+ (GTIN *)gtinWithGTINWithoutCheckDigit:(NSString *)gtinWithoutCheckDigit {
    NSString *gtin = [GTIN stringWithCheckDigitWithGTINWithoutCheckDigit:gtinWithoutCheckDigit];
    GTIN *g = [GTIN new];
    g.gtin = gtin;
    g.format = GTINFormatForLength(gtin.length);
    
    return g;
}

+ (BOOL)isValidString:(NSString *)gtin {
    return [GTIN isValidString:gtin forFormat:GTINFormatCount()];
}

+ (BOOL)isValidString:(NSString *)gtin forFormat:(GTINFormat)format {
    // Check format of barcode for validity
    if (![GTIN stringMatchesFormat:gtin]) {
        return false;
    }
    
    NSUInteger gtinLength = gtin.length;
    if (format != GTINFormatCount() && gtinLength != GTINFormatLength(format)) {
        return false;
    }
    
    int checkSum = 0;
    int weightBit = gtinLength % 2;
    for (int i = 0; i < gtinLength; i++) {
        int weight = i % 2 == weightBit ? 3 : 1;
        checkSum += [[gtin substringWithRange:NSMakeRange(i, 1)] intValue] * weight;
    }
    
    return checkSum % 10 == 0;
}

+ (BOOL)isValidString8:(NSString *)gtin {
    return [GTIN isValidString:gtin forFormat:GTINFormatGTIN_8];
}

+ (BOOL)isValidString12:(NSString *)gtin {
    return [GTIN isValidString:gtin forFormat:GTINFormatGTIN_12];
}

+ (BOOL)isValidString13:(NSString *)gtin {
    return [GTIN isValidString:gtin forFormat:GTINFormatGTIN_13];
}

+ (BOOL)isValidString14:(NSString *)gtin {
    return [GTIN isValidString:gtin forFormat:GTINFormatGTIN_14];
}

+ (int)calculateCheckDigitWithGTINWithoutCheckDigit:(NSString *)gtinWithoutCheckDigit {
    if (![GTIN stringMatchesFormat:gtinWithoutCheckDigit format:GTINFormatCount() offset:1]) {
        @throw [NSException exceptionWithName:GTINFormatException reason:[NSString stringWithFormat:@"String %@ is not a valid partial gtin", gtinWithoutCheckDigit] userInfo:nil];
    }
    
    int checkSum = 0;
    NSUInteger gtinLength = gtinWithoutCheckDigit.length;
    int weightBit = (gtinLength + 1) % 2;
    for (int i = 0; i < gtinLength; i++) {
        int weight = i % 2 == weightBit ? 3 : 1;
        checkSum += [[gtinWithoutCheckDigit substringWithRange:NSMakeRange(i, 1)] intValue] * weight;
    }
    
    return checkSum % 10 == 0 ? 0 : 10 - (checkSum % 10);
}

+ (NSString *)stringWithCheckDigitWithGTINWithoutCheckDigit:(NSString *)gtinWithoutCheckDigit {
    return [NSString stringWithFormat:@"%@%d", gtinWithoutCheckDigit,
            [GTIN calculateCheckDigitWithGTINWithoutCheckDigit:gtinWithoutCheckDigit]];
}

+ (BOOL)stringMatchesFormat:(NSString *)gtin {
    return [GTIN stringMatchesFormat:gtin format:GTINFormatCount() offset:0];
}

+ (BOOL)stringMatchesFormat:(NSString *)gtin format:(GTINFormat)format {
    return [GTIN stringMatchesFormat:gtin format:format offset:0];
}

+ (BOOL)stringMatchesFormat8:(NSString *)gtin {
    return [GTIN stringMatchesFormat:gtin format:GTINFormatGTIN_8 offset:0];
}

+ (BOOL)stringMatchesFormat12:(NSString *)gtin {
    return [GTIN stringMatchesFormat:gtin format:GTINFormatGTIN_12 offset:0];
}

+ (BOOL)stringMatchesFormat13:(NSString *)gtin {
    return [GTIN stringMatchesFormat:gtin format:GTINFormatGTIN_13 offset:0];
}

+ (BOOL)stringMatchesFormat14:(NSString *)gtin {
    return [GTIN stringMatchesFormat:gtin format:GTINFormatGTIN_14 offset:0];
}

+ (BOOL)stringMatchesFormat:(NSString *)gtin format:(GTINFormat)format offset:(NSUInteger)offset {
    if (gtin == nil) {
        @throw [NSException exceptionWithName:NSInvalidArgumentException reason:@"gtin is nil" userInfo:nil];
    }
    
    // Check length
    NSUInteger gtinLength = gtin.length;
    BOOL validLength = false;
    if (format != GTINFormatCount()) {
        validLength = gtinLength == GTINFormatLength(format);
    } else {
        for (id gtinFormatValue in GTINFormatValues()) {
            GTINFormat gtinFormat = (GTINFormat) [gtinFormatValue intValue];
            if (gtinLength == GTINFormatLength(gtinFormat) - offset) {
                validLength = true;
                break;
            }
        }
    }
    if (!validLength) {
        return false;
    }
    
    // Check whether is a number
    return [[NSPredicate predicateWithFormat:@"SELF MATCHES %@", @"^[0-9]+$"] evaluateWithObject:gtin];
}

- (GTINFormat)format {
    return _format;
}

- (NSUInteger)length {
    return GTINFormatLength(_format);
}

- (int)checkDigit {
    return [[_gtin substringFromIndex:_gtin.length - 1] intValue];
}

- (int)digitAtPosition:(NSUInteger)position {
    return [[_gtin substringWithRange:NSMakeRange(position, 1)] intValue];
}

- (NSString *)description {
    return _gtin;
}

- (BOOL)isEqual:(id)obj {
    return obj == self || ([obj isKindOfClass:[GTIN class]] && [((GTIN *) obj) isEqualToString:_gtin]);
}

- (BOOL)isEqualToGTIN:(GTIN *)gtin {
    return gtin == self || [gtin isEqualToString:_gtin];
}

- (BOOL)isEqualToString:(NSString *)gtin {
    return [_gtin isEqualToString:gtin];
}

- (NSUInteger)hash {
    return 37 + _gtin.hash;
}

- (void)encodeWithCoder:(NSCoder *)encoder {
    [encoder encodeObject:_gtin forKey:kGtinKey];
}

#ifdef DEBUG
+ (void)load {
    [[NSUserDefaults standardUserDefaults] setValue:@"XCTestLog,GcovTestObserver"
                                             forKey:@"XCTestObserverClass"];
}
#endif

@end
